### DISCUSSION TITLE
Run PipelineC only once for given pipeline stages.

### DISCUSSION THREAD

**Discussion by bartokon** [JulianKemmerer/PipelineC#40] [Ideas]
Hi, is it possible to run compilation of some file for given pipeline stages? (not main_mhz pragma just one time run)
for example ./pipelineC blink.c -start 10 -stop 10 (10 pipeline stages) and return frequency to some file.

I would like to try some machine learning algorithms and see if they could do some magic with pipelining :)

**Comment by JulianKemmerer** [comment]
Hey there @bartokon ! thanks for reaching out

How does this sound, I just checked in a little example pipeline [examples/pipeline.c](https://github.com/JulianKemmerer/PipelineC/blob/master/examples/pipeline.c). Pull the repo.

It has a synthesis/`PART` tool set, but not `MAIN_MHZ`.

The default behavior of the tool
`./src/pipelinec ./examples/pipeline.c` is to see that there is no specified frequency and start from 0 clocks (1 pipeline stage), then try 1 clock, 2 clock, etc. Almost what you want.

Supplying the `--start` flag gets closer:
`./src/pipelinec ./examples/pipeline.c --start 9 --sweep` will make the tool `start` from a 9 clock (10 stage, no IO regs) pipeline and `sweep` upwards trying 10 clock, 11 clock, etc.

There is no `--stop` right now. Normally the tool stops when it meets timing. But in this case its kinda just up to the user to see what the timing results are as it climbs up in pipeline depth and kill the program, examine the logs manually (since there is no known timing goal).

Does any of this help? What would you like a potential new `--stop` to do - prepare to run/continue with whatever it worked on last before stopping?

Happy Friday!

Take care,
Julian

**Reply by bartokon** [reply]
Hello Julian,

Let me tell about what I would like to experiment on. 
First let's take for example that we have "./src/pipelinec ./examples/pipeline.c --start 9 --sweep --once" --> start from 9clock and stop there do not try to opt anymore. PipelineC will save somewhere results frequency and latency (if it meets timing or failed). 

Now let's see what would happen if we ran this thread like 10 times
"./src/pipelinec ./examples/pipeline.c --start 4 --sweep --once"
"./src/pipelinec ./examples/pipeline.c --start 123 --sweep --once"
"./src/pipelinec ./examples/pipeline.c --start 50 --sweep --once"
etc... 

We would have some samples to work on. 
Now we could: 
1. Predict the Polynomial that supposedly presents how freq and latency scales with "--start N" command, and approximate more points to have better accuracy. Determinate the limit of passing time check.
2. Use genetic algorithm to evolve and find best "optimal" --start N and Freq pair. 
3. Use two GAN's that would fight with each other and generate better solution → less latency and more freq.

This would be fun to do as an extra "overlay" application for PipelineC.

What do you think about it? I know it is not optimal calculation wise, as you need some samples to work on.

Take care, 
Have a great weekend.

**Reply by JulianKemmerer** [reply]
OK cool - I believe I understand what you are getting at. Predicting how freq and latency scales sounds very useful. Anything to have the tool do better guesses / i.e. fewer synthesis/pnr runs would be great to have. And interesting to think about the genetic and GAN approaches with the tool.

I think it makes sense to have a `--stop N` argument for now (where N equals --start N, as opposed to subset case of one run `--once` at the moment), in case people do want a few results / want to let the tool try to make meet timing a bit. :+1:  seems very useful to say - at this `stop` point just give up, report results to user.

I dont have any experience with the machine learning side - but happy to get `--stop` implemented for you pretty quickly I believe. Can proceed from there. Ttys :)

**Comment by JulianKemmerer** [comment]
Righto @bartokon. Pull the repo again.

I implemented a `--stop N`. (see `-h`)

Running 
`./src/pipelinec ./examples/pipeline.c --start 9 --stop 9`

You will see something like

```
$ ./src/pipelinec ./examples/pipeline.c --start 9 --stop 9

██████╗ ██╗██████╗ ███████╗██╗     ██╗███╗   ██╗███████╗ ██████╗
██╔══██╗██║██╔══██╗██╔════╝██║     ██║████╗  ██║██╔════╝██╔════╝
██████╔╝██║██████╔╝█████╗  ██║     ██║██╔██╗ ██║█████╗  ██║     
██╔═══╝ ██║██╔═══╝ ██╔══╝  ██║     ██║██║╚██╗██║██╔══╝  ██║     
██║     ██║██║     ███████╗███████╗██║██║ ╚████║███████╗╚██████╗
╚═╝     ╚═╝╚═╝     ╚══════╝╚══════╝╚═╝╚═╝  ╚═══╝╚══════╝ ╚═════╝

Output directory: /media/1TB/Dropbox/PipelineC/git/PipelineC/pipelinec_output_pipeline.c_18887
================== Parsing C Code to Logical Hierarchy ================================
Parsing: /media/1TB/Dropbox/PipelineC/git/PipelineC/examples/pipeline.c
Preprocessing file...
Parsing C syntax...
Parsing non-function definitions...
Parsing derived fsm logic functions...
Doing old-style code generation based on PipelineC supported text patterns...
Parsing function logic...
Using VIVADO synthesizing for part: xc7a35ticsg324-1l
Parsing function: my_pipeline
Elaborating pipeline hierarchies down to raw HDL logic...
Doing obvious logic trimming/collapsing...
Writing generated PipelineC code from elaboration to output directories...
Writing cache of parsed information to file...
================== Writing Resulting Logic to File ================================
Building map of combinatorial logic...
Writing report of module instances...
Writing VHDL files for all functions (as combinatorial logic)...
Writing func my_pipeline ...
Writing func BIN_OP_GT_uint1_t_uint64_t_uint64_t ...
Writing func MUX_uint64_t_uint1_t_uint64_t_uint64_t ...
Writing func BIN_OP_PLUS_uint65_t_uint64_t_uint64_t ...
Writing func BIN_OP_INFERRED_MULT_uint128_t_uint64_t_uint64_t ...
Writing multi main top level files...
Writing the constant struct+enum definitions as defined from C code...
Writing clock cross definitions as parsed from C code...
Writing finalized comb. logic synthesis tool files...
Warning: No frequency associated with clock clk_None. Missing MAIN_MHZ?
Output VHDL files: /media/1TB/Dropbox/PipelineC/git/PipelineC/pipelinec_output_pipeline.c_18887/read_vhdl.tcl
================== Adding Timing Information from Synthesis Tool ================================
Synthesizing as combinatorial logic to get total logic delay...

Function: BIN_OP_GT_uint1_t_uint64_t_uint64_t Cached path delay(ns): 3.267
Function: MUX_uint64_t_uint1_t_uint64_t_uint64_t Cached path delay(ns): 1.643
Function: BIN_OP_PLUS_uint65_t_uint64_t_uint64_t Cached path delay(ns): 3.928
Function: BIN_OP_INFERRED_MULT_uint128_t_uint64_t_uint64_t Cached path delay(ns): 14.978
Synthesizing function: my_pipeline
...Waiting on synthesis for: my_pipeline
Running: /media/1TB/Programs/Linux/Xilinx/Vivado/2019.2/bin/vivado -log /media/1TB/Dropbox/PipelineC/git/PipelineC/pipelinec_output_pipeline.c_18887/examples/pipeline.c/my_pipeline/vivado_0CLK_6f2c5aad.log -source "/media/1TB/Dropbox/PipelineC/git/PipelineC/pipelinec_output_pipeline.c_18887/examples/pipeline.c/my_pipeline/my_pipeline_0CLK_6f2c5aad.syn.tcl" -journal /media/1TB/Dropbox/PipelineC/git/PipelineC/pipelinec_output_pipeline.c_18887/examples/pipeline.c/my_pipeline/vivado.jou -mode batch
my_pipeline Path delay (ns): 14.166 = 70.59155724975292 MHz

================== Beginning Throughput Sweep ================================
Function: my_pipeline Target MHz: None
Setting all instances to comb. logic to start...
Starting with blank sweep state...
Switching to coarse sweep since only main function has no specified frequency...
Doing coarse sweep only...
Main Function: my_pipeline does not have a set target frequency.
Building timing params up from comb. logic...
my_pipeline : sliced coarsely ~= 9 clocks latency...
Updating output files...
Running syn w slices...
Running: /media/1TB/Programs/Linux/Xilinx/Vivado/2019.2/bin/vivado -log /media/1TB/Dropbox/PipelineC/git/PipelineC/pipelinec_output_pipeline.c_18887/examples/pipeline.c/my_pipeline/vivado_9CLK_5062cd6a.log -source "/media/1TB/Dropbox/PipelineC/git/PipelineC/pipelinec_output_pipeline.c_18887/examples/pipeline.c/my_pipeline/my_pipeline_9CLK_5062cd6a.syn.tcl" -journal /media/1TB/Dropbox/PipelineC/git/PipelineC/pipelinec_output_pipeline.c_18887/examples/pipeline.c/my_pipeline/vivado.jou -mode batch
my_pipeline Clock Goal: 1000.00 (MHz) Current: 253.16 (MHz)(3.95 ns) 9 clks
Stopping at 9 clocks latency...
================== Writing Results of Throughput Sweep ================================
Warning: No frequency associated with clock clk_1000. Missing MAIN_MHZ?
Output VHDL files: /media/1TB/Dropbox/PipelineC/git/PipelineC/pipelinec_output_pipeline.c_18887/read_vhdl.tcl
Done.
```

I think you definitely want 
`my_pipeline Clock Goal: 1000.00 (MHz) Current: 253.16 (MHz)(3.95 ns) 9 clks`
Saying that in this 9 clocks (9 registers, 10 comb logic stages) had an FMAX of `253.16 (MHz)` for the demo `my_pipeline` C function.
Just a note: There is some subtlety here in running the tool set to 1000MHz as the 'try your hardest' setting - some tools give up early if the goal is too high. But it has at least been enough of an indicator for my work/this tool/etc.

Also, this was using Vivado for an Artix 7 35T. You can see logs an mentioned in the output
`Running: /media/1TB/Programs/Linux/Xilinx/Vivado/2019.2/bin/vivado -log /media/1TB/Dropbox/PipelineC/git/PipelineC/pipelinec_output_pipeline.c_18887/examples/pipeline.c/my_pipeline/vivado_9CLK_5062cd6a.log`
That has resource utilization and timing reports inside.

Let me know if this is helpful eh!

Take care!

**Comment by bartokon** [comment]
Hi Julian, 

I have a few simple requests :)

- First, I need more knobs to work on. Now PipelineC api supports N stages only which is 1 input problem and some ML don't work well on these. Could you provide me the way to access more parameters? basically ML would try to guess params for optimal pipelining.
I saw some parameters in syn.py are these important for pipelining? Could I pass these parameters by specifying some --parameters file?

- Right now I'm getting stages and frequency by dumping pipelineC output to "N_latency.txt" and searching for "my_pipeline Clock Goal: 1000.00 (MHz) Current: Y (MHz)(3.95 ns) N clks" line. Could you dump info to some file like if it passed timing, current freq, latency parameters? manually searching is just sub optimal and maybe you have already that data extracted in your code?

Do you think that I could use another synthesis tool for faster architecture estimation than Vivado? Just to find optimal starting point for algorithms? Maybe Yosys?

Take care!

**Reply by JulianKemmerer** [reply]
Ooo alright so some notes:

The base data structure at the core of everything right now is
https://github.com/JulianKemmerer/PipelineC/blob/f422157d34db1f5512b4178c5d304ec9b5f273f7/src/SYN.py#L111
Which, for each module instance, mainly has:
* `slices` : list of floats , a pipeline split by two registers into 3 equal stages would be [0.333, 0.333] 
*  `input|output_regs`: flag to put input or output regs in, adding cycle or two of latency to module
* `params_are_fixed`: flag for this logic cant be autopipelined/dont touch params

Right now the coarse `--start N` single 'N clock' pipeline requires trying to divide the single (not setup for multiple) top level pipeline down to the lowest level `raw vhdl` submodules which need `slices` in particular places.

Alternatively, and by default for multi-main-function designs, the tool builds pipelines not by saying 'make it N clocks total' but something like 'make all multipliers take 3 clocks - then tell me the total latency of the design'  (note the 'make all multipliers take 3 clocks' _does_ look exactly like the alternative case of 'make it N clocks total', but just for a submodule instead of top level)

Let me know if that gives you any ideas on what params and how they might be exposed to you - ill respond to your second bullet soon

**Reply by JulianKemmerer** [reply]
Also the parameters you found in the land of magic numbers
https://github.com/JulianKemmerer/PipelineC/blob/f422157d34db1f5512b4178c5d304ec9b5f273f7/src/SYN.py#L30

Are for if you let the PipelineC tool do its own sweep (instead of start=stop one run). Kinda hacked together heuristics and such. Ideally replaceable by ML in this journey as well.

**Reply by JulianKemmerer** [reply]
Re : `searching for` text in an output log. 
I am going to create an issue. How should PipelineC package up results for use in other tools - ex. csv, json or something idk? https://github.com/JulianKemmerer/PipelineC/issues/41

But for now, and especially since you are only doing a single start=stop run - there will only be one line that has `Clock Goal: 1000.00 (MHz) Current:` in it - so doing a bunch of runs in parallel (if you can - will get to slow vivado) - and then mass grepping the logs would be the easiest way to get your numbers right now.

**Reply by JulianKemmerer** [reply]
Re: `Yosys` 
Absolutely. To be clear, that flow to get timing information is `GHDL->Yosys->nextpnr`.

This page has info on how to easily setup Yosys,etc :) https://github.com/JulianKemmerer/PipelineC/wiki/Running-the-Tool

And then specifying a supported `#pragma PART` FPGA part - i.e. right now only ECP5(U), see main.c for example part numbers, will make the tool use that yosys flow by default (as opposed to Lattice tools).

**Reply by bartokon** [reply]
Okay, I will wait for that "output" CSV (or other) implementation and I will try that Yosysy flow. 
That output file should have parameters listed and synth result like latency and max freq for used latency.  

There could be also input CSV file that contains parameters for pipelineC. That way, the interface between tools should be pretty easy to do. PipelineC could read some (or all) parameters like slice factor, cycles for operation, io regs etc. the more variables to tune for ML the funnier it gets :D

**Reply by JulianKemmerer** [reply]
Hmm these are great points regarding both taking in some, ex. csv file pipeline description
And also being able to output some pipeline description file

I suppose both will be necessary when you come back to me with the bunch of params the NN churned out to try and run synthesis on to see if it worked :)

**Reply by JulianKemmerer** [reply]
I suppose I can dump entire configuration, that is
` for each module instance, mainly has: slices ...input|output_regs`

Something like
`inst_name, slices, in_reg, out_reg` in a csv ... can be many thousands of lines but you wanted knobs :control_knobs:

**Reply by bartokon** [reply]
Haha, can be even thousands parameters! The one and only thing is that it needs to be easy to manipulate from the "top". 

For example, we could have some hidden function that describes Frequency. 
Y = MIN(X1 + X2 + X3). So the worst freq depends on Xn instance. 
X1 = Z1*param1+Z2/param2..... etc. ML need to guess optimal param1 and param2 to get the best freq for given instance, find the correlation between arguments and guess better and better with more accuracy. 
The ML model can be generated for N input parameters, number of parameters could be extracted from combinatory synthesis (nb of function = nb of instances = X1, X2, X3).

Julian, just don't overwork yourself, start with what is easiest for you to implement, just in proper easy to manipulate format ;D
The model will take all the parameters as an input data set, so they even can be in one CSV file, and predict an optimal value.

**Reply by JulianKemmerer** [reply]
Csv is indeed seeming reasonable :+1:

**Comment by bartokon** [comment]
I have some errors on Yosys GHDL,
```
Output directory: /home/bartek/PipelineC/elo4
================== Parsing C Code to Logical Hierarchy ================================
Parsing: /home/bartek/PipelineC/examples/pipeline2.c
Preprocessing file...
Parsing C syntax...
Parsing non-function definitions...
Parsing derived fsm logic functions...
Doing old-style code generation based on PipelineC supported text patterns...
Parsing function logic...
Using OPEN_TOOLS synthesizing for part: LFE5UM5G-85F-8BG756C
Parsing function: my_pipeline
Elaborating pipeline hierarchies down to raw HDL logic...
Doing obvious logic trimming/collapsing...
Writing generated PipelineC code from elaboration to output directories...
Writing cache of parsed information to file...
================== Writing Resulting Logic to File ================================
Building map of combinatorial logic...
Writing report of module instances...
Writing VHDL files for all functions (as combinatorial logic)...
Writing func my_pipeline ...
Writing func BIN_OP_GT_uint1_t_uint64_t_uint64_t ...
Writing func MUX_uint64_t_uint1_t_uint64_t_uint64_t ...
Writing func BIN_OP_PLUS_uint65_t_uint64_t_uint64_t ...
Writing func BIN_OP_INFERRED_MULT_uint128_t_uint64_t_uint64_t ...
Writing multi main top level files...
Writing the constant struct+enum definitions as defined from C code...
Writing clock cross definitions as parsed from C code...
Writing finalized comb. logic synthesis tool files...
Output VHDL files: /home/bartek/PipelineC/elo4/vhdl_files.txt
================== Adding Timing Information from Synthesis Tool ================================
Synthesizing as combinatorial logic to get total logic delay...

Function: MUX_uint64_t_uint1_t_uint64_t_uint64_t Cached path delay(ns): 1.2429926290537097
Synthesizing function: BIN_OP_GT_uint1_t_uint64_t_uint64_t
Synthesizing function: BIN_OP_PLUS_uint65_t_uint64_t_uint64_t
Synthesizing function: BIN_OP_INFERRED_MULT_uint128_t_uint64_t_uint64_t
...Waiting on synthesis for: BIN_OP_GT_uint1_t_uint64_t_uint64_t
Running: /home/bartek/PipelineC/elo4/built_in/BIN_OP_PLUS_uint65_t_uint64_t_uint64_t/BIN_OP_PLUS_uint65_t_uint64_t_uint64_t_0CLK_de264c78_top.sh
Running: /home/bartek/PipelineC/elo4/built_in/BIN_OP_INFERRED_MULT_uint128_t_uint64_t_uint64_t/BIN_OP_INFERRED_MULT_uint128_t_uint64_t_uint64_t_0CLK_de264c78_top.sh
Running: /home/bartek/PipelineC/elo4/built_in/BIN_OP_GT_uint1_t_uint64_t_uint64_t/BIN_OP_GT_uint1_t_uint64_t_uint64_t_0CLK_de264c78_top.sh
Command failed:
bash BIN_OP_INFERRED_MULT_uint128_t_uint64_t_uint64_t_0CLK_de264c78_top.sh


Command failed:
bash BIN_OP_GT_uint1_t_uint64_t_uint64_t_0CLK_de264c78_top.sh


Command failed:
bash BIN_OP_PLUS_uint65_t_uint64_t_uint64_t_0CLK_de264c78_top.sh


Traceback (most recent call last):
  File "/home/bartek/PipelineC/src/pipelinec", line 213, in <module>
    parser_state = SYN.ADD_PATH_DELAY_TO_LOOKUP(parser_state)
  File "/home/bartek/PipelineC/src/SYN.py", line 2845, in ADD_PATH_DELAY_TO_LOOKUP
    parsed_timing_report =  my_async_result.get()
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 771, in get
    raise self._value
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/bartek/PipelineC/src/OPEN_TOOLS.py", line 242, in SYN_AND_REPORT_TIMING
    return SYN_AND_REPORT_TIMING_NEW(parser_state, multimain_timing_params, inst_name, total_latency, hash_ext, use_existing_log_file)
  File "/home/bartek/PipelineC/src/OPEN_TOOLS.py", line 344, in SYN_AND_REPORT_TIMING_NEW
    C_TO_LOGIC.GET_SHELL_CMD_OUTPUT(syn_imp_bash_cmd, cwd=output_directory)
  File "/home/bartek/PipelineC/src/C_TO_LOGIC.py", line 251, in GET_SHELL_CMD_OUTPUT
    raise Exception("Command failed!")
Exception: Command failed!

```
and
```
` /----------------------------------------------------------------------------\
 |                                                                            |
 |  yosys -- Yosys Open SYnthesis Suite                                       |
 |                                                                            |
 |  Copyright (C) 2012 - 2020  Claire Xenia Wolf <claire@yosyshq.com>         |
 |                                                                            |
 |  Permission to use, copy, modify, and/or distribute this software for any  |
 |  purpose with or without fee is hereby granted, provided that the above    |
 |  copyright notice and this permission notice appear in all copies.         |
 |                                                                            |
 |  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES  |
 |  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF          |
 |  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR   |
 |  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES    |
 |  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN     |
 |  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF   |
 |  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.            |
 |                                                                            |
 \----------------------------------------------------------------------------/

 Yosys 0.11+7 (git sha1 d39d4e11d, clang 10.0.0-4ubuntu1 -fPIC -Os)


-- Running command `ghdl /home/bartek/PipelineC/elo4/c_structs_pkg.pkg.vhd /home/bartek/PipelineC/elo4/built_in/BIN_OP_INFERRED_MULT_uint128_t_uint64_t_uint64_t/BIN_OP_INFERRED_MULT_uint128_t_uint64_t_uint64_t_0CLK_de264c78_top.vhd /home/bartek/PipelineC/elo4/built_in/BIN_OP_INFERRED_MULT_uint128_t_uint64_t_uint64_t/BIN_OP_INFERRED_MULT_uint128_t_uint64_t_uint64_t_0CLK_de264c78.vhd  -e BIN_OP_INFERRED_MULT_uint128_t_uint64_t_uint64_t_0CLK_de264c78_top; synth_ecp5 -top BIN_OP_INFERRED_MULT_uint128_t_uint64_t_uint64_t_0CLK_de264c78_top -json BIN_OP_INFERRED_MULT_uint128_t_uint64_t_uint64_t_0CLK_de264c78_top.json' --

1. Executing GHDL.
/home/bartek/PipelineC/elo4/built_in/BIN_OP_INFERRED_MULT_uint128_t_uint64_t_uint64_t/BIN_OP_INFERRED_MULT_uint128_t_uint64_t_uint64_t_0CLK_de264c78.vhd:11:10: unit "float_pkg" not found in library "ieee"
use ieee.float_pkg.all;
         ^
ERROR: vhdl import failed.
Warning: Use of default value for --package is deprecated. Please add '--package CABGA381' to arguments.
ERROR: Failed to open JSON file 'BIN_OP_INFERRED_MULT_uint128_t_uint64_t_uint64_t_0CLK_de264c78_top.json'.
1 warning, 1 error`
```

**Reply by JulianKemmerer** [reply]
@bartokon this was a dumb issue with GHDL not being passed its 'use VHDL 2008' flag. It has been fixed. Pull repo.

**Reply by bartokon** [reply]
Thanks, probably with Yosys I will iterate much faster than before.

**Comment by bartokon** [comment]
Hey Julian, 
How many sweep passes do you need to find max freq for example programs (for instance mypipeline?) and what is that freq for Artix7?

For now I have created simple decision tree and these are results...
Manual gathered data + some decision tree tests ->

![manual](https://user-images.githubusercontent.com/47795888/140809728-bcfb017e-4796-46ea-85c1-f7d2347edcfe.png)

This is full auto ->


How does it works? It probes "0" and "MAX_LATENCY" that is for now 1024 and do rest on it's own. It guesses max freq in range(1, MAX_LATENCY) and output is rounded and passed to Vivado synth. This is output for 10 iterations
![auto](https://user-images.githubusercontent.com/47795888/140810241-f7a0a628-ad33-467c-afa9-f1d60d64d498.png)
and this is console output.
`Running pipeline for 1024 cycles
[1]
[337.61 337.61 337.61 ... 337.61 337.61 337.61]
Running pipeline for 1 cycles
[513]
[ 95.86  95.86  95.86 ... 337.61 337.61 337.61]
Running pipeline for 513 cycles
[258]
[ 95.86  95.86  95.86 ... 337.61 337.61 337.61]
Running pipeline for 258 cycles
[130]
[ 95.86  95.86  95.86 ... 337.61 337.61 337.61]
Running pipeline for 130 cycles
[195]
[ 95.86  95.86  95.86 ... 337.61 337.61 337.61]
Running pipeline for 195 cycles
[227]
[ 95.86  95.86  95.86 ... 337.61 337.61 337.61]
Running pipeline for 227 cycles
[212]
[ 95.86  95.86  95.86 ... 337.61 337.61 337.61]
Running pipeline for 212 cycles
[220]
[ 95.86  95.86  95.86 ... 337.61 337.61 337.61]
Running pipeline for 220 cycles
[224]
[ 95.86  95.86  95.86 ... 337.61 337.61 337.61]
Running pipeline for 224 cycles
[226]
[ 95.86  95.86  95.86 ... 337.61 337.61 337.61]
Running pipeline for 226 cycles`
and this is decision tree...
![funny_graph](https://user-images.githubusercontent.com/47795888/140810375-78bd9a21-9c1e-441a-a71c-c2049a7f9f9d.png)
 
I need to test it on more programs... but it seems to work on this example >:3
Basically according to the decision tree it goes something like this:
if latency <=65.5 True: you will get ~95Mhz 
False: is latency <=385.5? True: you will get ~337Mhz 
False: is latency <=162.5? True you will get ~412Mhz

**Reply by JulianKemmerer** [reply]
Ooo neat work doing binary search to look for fmax.

`how many sweep passes do you need to find max freq for example programs (for instance mypipeline?) and what is that freq for Artix7?`

Typically I am not trying to find the fmax for a given pipeline but just setting the tool to try and make a pipeline meet timing near my device's fmax. The fmax for pipelined designs is generally determined/limited by the FPGA technology (ex. Artix 7 FPGA clock tree limits all designs to like ~400MHz _at max_ in practice). So all maximally pipelined pipelines (if the pipelinec tool can produce such designs) have an FMAX around ~400MHz for this device.

However, it also important to note that targeting the absolute FMAX of your FPGA device for all modules will likely not work/is wasteful. In a real-world design you can probably expect 300-400MHz on an Artix 7 to be 'pretty good'.  (having deeper pipeline that _could_ reach fmax but isnt being used at fmax is just wasted latency/clock cycles).

That said - it depends on the size of the design. I mentioned some hand crafted heuristic type stuff I have setup, its easier to get that right for smaller designs. Larger designs take longer-per-iteration of synthesis/pnr and more iterations because the timing feedback I get reflects smaller and smaller parts of the entire design. Its a slow incremental process. It can be anywhere from a handful of synthesis runs, taking a few mins, to hundreds, taking multiple days. Lots of variables to consider.

What cha think eh?

**Reply by JulianKemmerer** [reply]
Also the tool does two rounds of synthesis/pnr
First doing combintorial logic to get logic+wire delays for all functions.
Then iteratively trying out its guesses pipeline params and getting feedback from syn+pnr tool

**Comment by JulianKemmerer** [comment]
Generally getting at what you want to `learn` here

Are you planning to use the C code as input to algorithm?

otherwise I suppose you are starting from something like the 
`inst_name, slices, in_reg, out_reg` in a csv I mentioned

Just a big list of function names/hierarchy structure - and params (not visible the original C code)

So some kind of map of learning

instance params = f(target freq, instance names)

After being trained with

fmax = f(instance names, instance params)

?

**Reply by bartokon** [reply]
I'm aiming to get better quality of results. 

For example if we set target for example 300+-10Mhz tool should find the lowest amount of pipeline stages (to limit resource usage) for given frequency, and that means "guide" pipelineC for optimal architecture, for example by changing CSV parameters.
From what I see, the ML part takes near 0 time compared to synthesis at this moment.
Now the goal is to find the fastest synth tool to iterate for general architecture then perform target synthesis in slow precise tool, like Vivado. If "Fast" synthesis results are just near "slow" precise results, general architecture should match thus it is like transfer synthesis learning 🗡️

I don't want to dig deep into inner workings of PipelineC just try to guide it to be a better tool with some ML models.
Right now, model is just trained with input data as latency and predict frequency for given .c file. Model predicts frequency output in some latency range and guess what latency for function is the fastest.
If ML can optimize one function, it could run multiple ML models to opt a few functions simultaneously, for example on GPU thus CPU could synth more functions in meantime. 

TL;DR
Want to learn parameters that describe optimal architecture. Lowest resources in target frequency.

**Reply by bartokon** [reply]
This is not a problem for ML model to evolve if you change a number of input parameters after some git commit in CSV, so I think this is a more flexible solution than hard-coded parameters.

**Reply by JulianKemmerer** [reply]
OK getting at your `Right now, model is just trained with input data as latency and predict frequency for given .c file`

So you want run the PipelinecC tool+syn+pnr, etc  to evaluate+learn `f`
`freq = f(latency, c code)`

The parameters I was planning to put into a csv are the `deep into inner workings` I dont think you want to get into.
And if the only info/params you need is freq and total latency, I dont think we need a one line, two or three number csv file output for that right? But if its really a friction point let me know...

Thinking that it doesnt seem like there is anything stopping you from doing that^ experiment as is right now? Thinking about how you mentioned wanting to wait for csv before proceeding?

Let me know whats up

Take care!

**Reply by bartokon** [reply]
"deep into inner workings" I thought about parsing C into Verilog and adding pipeline regs manually. I want pipelineC to do it (with pipeline parameters generated by ML). Do not, don't reinvent the wheel twice!
I'm calling pipelineC by using system calls, and it is giving me some output data for parameter that I provide, for example --start --stop latency. l just want more parameters to work with ;).

"Thinking that it doesn't seem like there is anything stopping you from doing that^ experiment as is right now? " yea I'm working with latency only at this moment, but in the future I would like to test it with more control about code generation, that is control "inst_name, slices, in_reg, out_reg". 

So I want to "trying out its guesses pipeline params and getting feedback from syn+pnr tool"

Take care, I will have some more fun with PipelineC today as well 👍

**Comment by JulianKemmerer** [comment]
`--coarse --start N --stop N` builds a single fixed pipeline depth, and the sweep records each iteration's result in <out_dir>/<top>/sweep_history.json (see docs/SWEEP_DESIGN.md).
